### PR TITLE
WIP: Adding terms interface - do not merge

### PR DESCRIPTION
**Pull request by dfm**
I'm adding a terms interface like in the Python code. It's not complete but you can do something like the following:

```julia
include("src/terms.jl")

t1 = RealTerm(0.5, 0.1)
t2 = ComplexTerm(0.1, 0.5, 1.0, -0.5)  # All the parameters are _log_
t3 = SHOTerm(0.5, 1.5, 0.3)

get_all_coefficients(t1)  # Returns a_real, c_real, a_complex, b_complex, c_complex, d_complex arrays

get_parameter_vector(t2)

kernel = t1 + t2 * t3
```

etc.

### CODE REVIEW

**Comment by dfm** [comment]
So the syntax with my proposed method is something like:

```julia
include("src/gp.jl")

kernel = RealTerm(0.5, 1.0) + SHOTerm(0.1, 2.0, -0.5)
gp = Celerite(kernel)

x = sort(10*rand(500))
y = sin(x)
yerr = 0.01

compute(gp, x, yerr)
println(log_likelihood(gp, y))
```

This hasn't been very well tested but this is as close as I could get to the Python syntax. What do you think?

**Comment by ericagol** [comment]
That looks pretty cool, but I'm not familiar enough with Julia objects 
to suggest an alternative. Maybe Will could advise?

**Comment by dfm** [comment]
Sure. Let's ping him: @farr

BTW if you check out my branch, this code does work.

**Comment by ericagol** [comment]
> Sure. Let's ping him: @farr <https://github.com/farr>
>
> BTW if you check out my branch, this code does work.
>
Thanks... not sure how to do that, but I won't merge until you
give me the goahead.

**Comment by dfm** [comment]
He's been notified but he's also probably asleep. We'll see what he says tomorrow.

**Comment by dfm** [comment]
I've written a "tutorial" with my new interface and posted it here: http://celerite.readthedocs.io/en/latest/tutorials/julia-first/

It's just a copy of the "Python: First Steps" tutorial but I haven't added any text yet.

**Comment by ericagol** [comment]
Cool.  If you're in tomorrow, maybe you can show me what you've done so 
far.

**Comment by farr** [comment]
Dan's proposal is also how I would implement this in Julia.  If you've got a natural "object" and want to define "methods" on it (e.g. stuff you would do in Python or C++ via object.method(arg1, arg2, ...)), then it makes most sense in Julia to do method(object, arg1, arg2, ...), just as here.

**Comment by dfm** [comment]
I'd be happy to have you merge this now if you like it. I've added some simple tests and this version is getting the right answers :-)

File: src/terms.jl [hunk 1]
```diff
@@ -0,0 +1,296 @@
...
+
+# Abstract base class and definitions
+abstract Term
+
+function get_terms(term::Term)
+    return tuple(term)
+end
+
+function get_real_coefficients(term::Term)
+    return (Array(Float64, 0), Array(Float64, 0))
```
Comment (ericagol) [inline src/terms.jl:11]: I'm confused about what this function does.  It seems like it returns a tuple of zero-length arrays?

File: src/terms.jl [hunk 2]
```diff
@@ -0,0 +1,296 @@
...
+end
+
+function get_complex_coefficients(term::Term)
+    return (Array(Float64, 0), Array(Float64, 0), Array(Float64, 0), Array(Float64, 0))
```
Comment (ericagol) [inline src/terms.jl:15]: Ditto for this function.

File: src/terms.jl (hunk 1 above)
Comment (dfm) [reply src/terms.jl:11]: That's correct. That's the default behavior for the abstract base class. If you look below you'll see the overloads for other types.

File: src/terms.jl [hunk 3]
```diff
@@ -0,0 +1,296 @@
...
+
+# A real term where b=0 and d=0
+type RealTerm <: Term
+    log_a::Float64
+    log_c::Float64
+end
+
+function get_real_coefficients(term::RealTerm)
+    return [exp(term.log_a)], [exp(term.log_c)]
+end
```
Comment (dfm) [inline src/terms.jl:205]: e.g. here ^

File: src/terms.jl [hunk 4]
```diff
@@ -0,0 +1,296 @@
...
+type ComplexTerm <: Term
+    log_a::Float64
+    log_b::Float64
+    log_c::Float64
+    log_d::Float64
+end
+
+function get_complex_coefficients(term::ComplexTerm)
+    return [exp(term.log_a)], [exp(term.log_b)], [exp(term.log_c)], [exp(term.log_d)]
+end
```
Comment (dfm) [inline src/terms.jl:230]: and here! ^

File: src/terms.jl (hunk 1 above)
Comment (dfm) [reply src/terms.jl:11]: The SHOTerm is where things really get fun!

File: src/terms.jl [hunk 5]
```diff
@@ -0,0 +1,296 @@
...
+        c_real = cat(1, c_real, coeffs[2])
+        a_complex = cat(1, a_complex, coeffs[3])
+        b_complex = cat(1, b_complex, coeffs[4])
+        c_complex = cat(1, c_complex, coeffs[5])
+        d_complex = cat(1, d_complex, coeffs[6])
+    end
+    return (a_real, c_real, a_complex, b_complex, c_complex, d_complex)
+end
+
+function +(t1::Term, t2::Term)
```
Comment (ericagol) [inline src/terms.jl:97]: Cool... so you are overloading the + sign to apply to sums of kernels?

Comment (dfm) [reply src/terms.jl:97]: yep! And the product (see below)!

**Comment by ericagol** [comment]
What happens if, say, you try to take a product of a real term
and an SHOterm?

**Comment by dfm** [comment]
What do you mean? The kernel function will be given by the product of the 2 and the code will use Eq (14) to solve it.

**Comment by dfm** [comment]
There's a bit of hairy logic in the product definition that should have some tests but I'm confident that it's correct because I copied it directly from the Python implementation and I've tested that like crazy.

**Comment by ericagol** [comment]
Yeah, that looks impressive!

Eric Agol
Astronomy Professor
University of Washington

**Pull request merged**